File: src/main/java/net/dean/jraw/ApiException.java
package net.dean.jraw;

/**
 * This class provides a way to translate errors returned by the Reddit API into Java Exceptions
 */
public class ApiException extends Exception {
    private final String reason;
    private final String explanation;

    /**
     * Instantiates a new ApiException
     *
     * @param reason The reason error string
     * @param explanation The localized explanation
     */
    public ApiException(String reason, String explanation) {
        super(String.format("API returned error: \"%s\" (\"%s\")", reason, explanation));
        this.reason = reason;
        this.explanation = explanation;
    }

    /**
     * Instantiates a new ApiException
     * @param message The reason this error was thrown
     * @param cause What caused the error
     */
    public ApiException(String message, Throwable cause) {
        super(message, cause);
        this.reason = null;
        this.explanation = null;
    }

    /**
     * Gets the reason error string. A full list can be found
     * <a href="https://github.com/reddit/reddit/blob/master/r2/r2/lib/errors.py">here</a>.
     *
     * @return The reason
     */
    public String getReason() {
        return reason;
    }

    /**
     * Gets the localized explanation
     * @return The explanation
     */
    public String getExplanation() {
        return explanation;
    }
}

File: lib/src/main/kotlin/net/dean/jraw/ApiException.kt
package net.dean.jraw

import com.fasterxml.jackson.databind.JsonNode

/**
 * Exception subclass for reddit's many errors
 *
 * See [here](https://github.com/reddit/reddit/blob/master/r2/r2/lib/errors.py) for a full list.
 */
class ApiException(val code: String, val explanation: String) : RuntimeException("API returned error: $code ($explanation)") {
    companion object {
        @JvmStatic fun from(node: JsonNode): ApiException {
            if (!node.isArray) throw IllegalArgumentException("JSON node must be an array")
            if (node.size() < 2) throw IllegalArgumentException("Expected at least 2 elements")
            return ApiException(node[0].asText(), node[1].asText())
        }
    }
}
